Based on https://github.com/ocaml/ocaml/commit/0f5046b8b48b7f6b9c343aab99d4b1314586802c.

--- origsrc/ocaml-4.01.0/Makefile	2013-06-17 08:15:18.000000000 -0500
+++ src/ocaml-4.01.0/Makefile	2016-06-05 17:58:20.000000000 -0500
@@ -761,8 +761,8 @@
 # Check that the stack limit is reasonable.
 
 checkstack:
-	@if $(BYTECC) -o tools/checkstack tools/checkstack.c; \
-	  then tools/checkstack; \
+	@if $(MKEXE) -o tools/checkstack$(EXE) tools/checkstack.c; \
+	  then tools/checkstack$(EXE); \
 	  else :; \
 	fi
 	@rm -f tools/checkstack
diff -ur origsrc/ocaml-4.01.0/configure src/ocaml-4.01.0/configure
--- origsrc/ocaml-4.01.0/configure	2013-08-23 01:22:36.000000000 -0500
+++ src/ocaml-4.01.0/configure	2016-06-05 17:55:48.000000000 -0500
@@ -323,10 +323,15 @@
   *,powerpc-*-aix*)
     bytecccompopts="-D_XOPEN_SOURCE=500";;
   gcc*,*-*-cygwin*)
+    case $host in
+      i[3456]86-*) flavor=cygwin;;
+      x86_64-*) flavor=cygwin64;;
+      *) err "Unknown Cygwin variant"
+    esac
     bytecccompopts="-fno-defer-pop $gcc_warnings -U_WIN32"
     dllccompopts="-U_WIN32 -DCAML_DLL"
     if test $withsharedlibs = yes; then
-      flexlink="flexlink -chain cygwin -merge-manifest -stack 16777216"
+      flexlink="flexlink -chain $flavor -merge-manifest -stack 16777216"
       flexdir=`$flexlink -where | dos2unix`
       if test -z "$flexdir"; then
         echo "flexlink not found: native shared libraries won't be available"
Only in src/ocaml-4.01.0: configure.orig
