Adapted from the ocaml-4.01.0-1 package in Cygwin.

--- origsrc/ocaml-4.01.0/byterun/Makefile	2013-03-28 11:10:24.000000000 -0500
+++ src/ocaml-4.01.0/byterun/Makefile	2016-06-05 10:54:18.000000000 -0500
@@ -21,7 +21,7 @@
 PICOBJS=$(OBJS:.o=.pic.o)
 
 SHARED_LIBS_TMP=$(SUPPORTS_SHARED_LIBRARIES:%false=)
-SHARED_LIBS_DEPS=$(SHARED_LIBS_TMP:%true=libcamlrun_shared.so)
+SHARED_LIBS_DEPS=$(SHARED_LIBS_TMP:%true=libcamlrun_shared.dll.a)
 
 all:: $(SHARED_LIBS_DEPS)
 
@@ -41,15 +41,16 @@
 	ar rc libcamlrund.a $(DOBJS)
 	$(RANLIB) libcamlrund.a
 
-libcamlrun_shared.so: $(PICOBJS)
-	$(MKDLL) -o libcamlrun_shared.so $(PICOBJS) $(BYTECCLIBS)
+libcamlrun_shared.dll.a: $(PICOBJS)
+	$(MKDLL) -o cygcamlrun_shared.dll -link -Wl,--out-implib,libcamlrun_shared.dll.a $(PICOBJS) $(BYTECCLIBS)
 
 install::
-	if test -f libcamlrun_shared.so; then \
-	  cp libcamlrun_shared.so $(LIBDIR)/libcamlrun_shared.so; fi
+	if test -f libcamlrun_shared.dll.a; then \
+	  cp cygcamlrun_shared.dll $(BINDIR); \
+	  cp libcamlrun_shared.dll.a $(LIBDIR)/libcamlrun_shared.dll.a; fi
 
 clean::
-	rm -f libcamlrun_shared.so
+	rm -f libcamlrun_shared.dll.a
 
 .SUFFIXES: .d.o .pic.o
 
